### ISSUE TITLE
Rare panic on socket reconnect in channel: concurrent `c.bindings` map iteration and map write

### ISSUE THREAD

**Issue by timadevelop** [nshafer/phx#5]
Sometimes, on reconnect, the following error occurs:

```
fatal error: concurrent map iteration and map write

goroutine 38 [running]:
github.com/nshafer/phx.(*Channel).trigger(0x0?, {0xc000486a70, 0x9}, 0x1b2, {0x68c540, 0xc0001e76e0})
        /home/vlad/go/pkg/mod/github.com/nshafer/phx@v0.2.2/channel.go:346 +0xe5
github.com/nshafer/phx.(*Channel).process(0xc0001dce10, 0xc000598500)
        /home/vlad/go/pkg/mod/github.com/nshafer/phx@v0.2.2/channel.go:335 +0x9f
github.com/nshafer/phx.(*Socket).onConnMessage(0xc0001a46e0, {0xc000112200?, 0xc000090190?, 0x0?})
        /home/vlad/go/pkg/mod/github.com/nshafer/phx@v0.2.2/socket.go:335 +0x3c5
github.com/nshafer/phx.(*Websocket).connectionReader(0xc0001dc240)
        /home/vlad/go/pkg/mod/github.com/nshafer/phx@v0.2.2/websocket.go:350 +0x88
created by github.com/nshafer/phx.(*Websocket).startup in goroutine 1
        /home/vlad/go/pkg/mod/github.com/nshafer/phx@v0.2.2/websocket.go:144 +0x205
```

I see https://github.com/nshafer/phx/commit/9a3b7aee718189ff4ebb590fff62f1f405bb0691, but I believe `RLock` is not enough in some cases.

Seems like `Lock` instead of a *recursive* lock needs to be used here:

https://github.com/nshafer/phx/blob/48d4ca840c510d43fc5cbbb1aa88c9570a5fe754/channel.go#L342-L350

because the callback goroutine (or other call that involves modifying bindings) might be concurrently executed on the same thread, which could modify the `c.bindings` map through `.Off` or other methods while iterating over the bindings on the same thread concurrently.

**Comment by timadevelop** [comment]
I'm not sure how others use phx.Channel, maybe they need reentrancy in some methods. Does it make sense to change all RLocks in channel.go to normal Locks?

Note that this affects all methods that operate on `c.bindings`

**Comment by nshafer** [comment]
So, `RLock` for a [RWMutex](https://pkg.go.dev/sync@go1.23.4#RWMutex) actually means Read lock, not recursive or re-entrant. But there was an issue elsewhere from the last commit where a read lock was being acquired for write operations, instead of a full lock. I have fixed and published 0.2.3. Thank you for letting me know, and for the stack trace that allowed me to figure it out. It's weird, I have had this in production for over a year, and have yet to have these panics for me. I guess I've just gotten really lucky. =P
